cli(debug.zip): upload profiles from node subdirectories

aa-joshi opened the pull request:
Previously, debug zip upload to Datadog would only considers and uploads cpu and
heap profiles which are in node sub-directories. This was inadequate because we
have cpu and heap profiles as part of `nodes/*/cpuprof` & `nodes/*/heapprof`
directories which will be ignored during upload. To address this, this patch
considers profiles of `cpuprof` and `heapprof` directories during the Datadog
upload. This patch also updates the profile explorer link generated after
upload.

Epic: [CRDB-52093](https://cockroachlabs.atlassian.net/browse/CRDB-52093)
Part of: CRDB-51119
Release note: None

-----
PFA screenshot after changes:
<img width="1330" alt="Screenshot 2025-07-09 at 11 09 58 AM" src="https://github.com/user-attachments/assets/c220e519-d55c-4a7b-9c7d-499199317e76" />

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/149315)
<!-- Reviewable:end -->

arjunmahishi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149315)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Abhinav1299 and @dhartunian)

___
*[`-- commits` line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149315#-OUId_e88M9-I6SjAjYA:-OUId_e88M9-I6SjAjYB:b9jjyaq):*
[This ticket](https://cockroachlabs.atlassian.net/browse/CRDB-51119) has a different issue right?

From my understanding :point_down: 

![image.png](https://files.reviewableusercontent.io/images/reviewable/11977524/nL1knIDMZlrzwnokAlPizE-M/image.png)


<!-- Sent from Reviewable.io -->

aa-joshi committed:
cli(debug.zip): upload profiles from node subdirectories

Previously, debug zip upload to Datadog would only considers and uploads cpu
and heap profiles which are in node sub-directories. This was inadequate
because we have cpu and heap profiles as part of `nodes/*/cpuprof` &
`nodes/*/heapprof` directories which will be ignored during upload. To address
this, this patch considers profiles of `cpuprof` and `heapprof` directories
during the Datadog upload. This patch also updates the profile explorer link
generated after upload.

Epic: CRDB-52093
Part of: CRDB-51119
Release note: None

aa-joshi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149315)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Abhinav1299, @arjunmahishi, and @dhartunian)

___
*[`-- commits` line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149315#-OUId_e88M9-I6SjAjYA:-OUiVJc-2lLfSBQV8iP3:b7rjvg4):*
<details><summary><i>Previously, arjunmahishi (Arjun Mahishi) wrote…</i></summary><blockquote>

[This ticket](https://cockroachlabs.atlassian.net/browse/CRDB-51119) has a different issue right?

From my understanding :point_down: 

![image.png](https://files.reviewableusercontent.io/images/reviewable/11977524/nL1knIDMZlrzwnokAlPizE-M/image.png)
</blockquote></details>

Yeah, it makes sense to include these. I have updated the PR to include the profiles from `cpuprof` and `heapprof`.


<!-- Sent from Reviewable.io -->

arjunmahishi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149315)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Abhinav1299 and @dhartunian)

___
*[`-- commits` line 4 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149315#-OVQbgSE2erLqG885r-L:-OVQbgSE2erLqG885r-M:undefined):*




_Suggestion:_
```
consider
```
___
*[`pkg/cli/zip_upload.go` line 346 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149315#-OVQdqaXD9omx8rBKVhg:-OVQdqaXD9omx8rBKVhh:b-pb00un) ([raw file](https://github.com/cockroachdb/cockroach/blob/69de0abed85e9a964439ff86bcf3d31f3ab94d9a/pkg/cli/zip_upload.go#L346)):*
> ```Go
> 		// and zippedHeapProfilePattern and try to extract the node ID from the suffix.
> 		nodeID := filepath.Base(filepath.Dir(path))
> 		if _, err := strconv.Atoi(nodeID); err != nil {
> ```

This might swallow actual errors or make it hard to debug issue with the path.

Is there a more deterministic way to distinguish between the two path patterns? Maybe with `strings.HasSuffix/HasPrefix` or even regex.
___
*[`pkg/cli/zip_upload.go` line 358 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149315#-OVQeY728KdqcY78Dbtz:-OVQeY728KdqcY78Dbu-:b68qvyg) ([raw file](https://github.com/cockroachdb/cockroach/blob/69de0abed85e9a964439ff86bcf3d31f3ab94d9a/pkg/cli/zip_upload.go#L358)):*
> ```Go
> 
> 	// start the upload pool
> 	for i := 0; i < noOfWorkers*100; i++ {
> ```

1. I think it's better to control the number noOfWorkers at the top where the variable is declared. This should be fixed in the other artifacts also (maybe in a different PR)
2. We don't have to do `* 100` right? In case of logs and table dumps, we are further patch the files before uploading. But here, since the number of requests made will always be <= `len(parts)`, we can just set a sane cap for the number of go routines we spin up. So, it can be 
    ```go
    noOfWorkers = min(debugZipUploadOpts.maxConcurrentUploads, len(parts), 1000) // or some other sane value
    ```
___
*[`pkg/cli/zip_upload.go` line 458 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149315#-OVQl55b768eqFvux55o:-OVQl55b768eqFvux55p:b-dt0bh0) ([raw file](https://github.com/cockroachdb/cockroach/blob/69de0abed85e9a964439ff86bcf3d31f3ab94d9a/pkg/cli/zip_upload.go#L458)):*
> ```Go
> 	fileName := filepath.Base(profilePath)
> 
> 	// Datadog only accepts CPU and heap profiles with filename as "cpu.pprof" or "heap.pprof".
> ```

lol, of course


<!-- Sent from Reviewable.io -->

arjunmahishi commented:
*[`pkg/cli/zip_upload.go` line 358 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149315#-OVQeY728KdqcY78Dbtz:-OVQlWX-BfU1Mq7A0dbv:bjwzjcb) ([raw file](https://github.com/cockroachdb/cockroach/blob/69de0abed85e9a964439ff86bcf3d31f3ab94d9a/pkg/cli/zip_upload.go#L358)):*
> further patch the files 

further batching the files*


<!-- Sent from Reviewable.io -->

aa-joshi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149315)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Abhinav1299, @arjunmahishi, and @dhartunian)

___
*[`pkg/cli/zip_upload.go` line 346 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149315#-OVQdqaXD9omx8rBKVhg:-OVRCIOK-sErz0OzYjoX:b7qz5i1) ([raw file](https://github.com/cockroachdb/cockroach/blob/69de0abed85e9a964439ff86bcf3d31f3ab94d9a/pkg/cli/zip_upload.go#L346)):*
<details><summary><i>Previously, arjunmahishi (Arjun Mahishi) wrote…</i></summary><blockquote>

This might swallow actual errors or make it hard to debug issue with the path.

Is there a more deterministic way to distinguish between the two path patterns? Maybe with `strings.HasSuffix/HasPrefix` or even regex.
</blockquote></details>

DONE. I have introduced a regex to differentiate between paths to fetch nodeID accordingly.
___
*[`pkg/cli/zip_upload.go` line 358 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149315#-OVQeY728KdqcY78Dbtz:-OVRCVBaBL_JTLKE1eKt:b-lnr9q3) ([raw file](https://github.com/cockroachdb/cockroach/blob/69de0abed85e9a964439ff86bcf3d31f3ab94d9a/pkg/cli/zip_upload.go#L358)):*
<details><summary><i>Previously, arjunmahishi (Arjun Mahishi) wrote…</i></summary><blockquote>

> further patch the files 

further batching the files*
</blockquote></details>

1. I agree. Let's address in the seperate PR as it would affect all uploads.
2. Yes, We are performing the upload per node. Now I have capped the workers to deterministic value of "max profiles associated with a node". Thanks for suggestion!
___
*[`-- commits` line 4 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149315#-OVQbgSE2erLqG885r-L:-OVRCH6c-oCHIcCnST19:b-896fix):*
Done.


<!-- Sent from Reviewable.io -->

arjunmahishi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149315#-:-OVRfGiD9q8tPlFl2n5q:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, and @dhartunian)



<!-- Sent from Reviewable.io -->

aa-joshi commented:
TFTR!
bors r+
